File: labs/lab10/lab10_4.cpp
#include <iostream>

class Handler {
protected:
    Handler* next;
public:
    Handler() : next(nullptr) {}
    virtual ~Handler() {}
    void setNext(Handler* next) {
        this->next = next;
    }
    virtual void handle(int request) {
        if (next) {
            next->handle(request);
        }
    }
};

class ConcreteHandler1 : public Handler {
public:
    void handle(int request) override {
        if (request == 1) {
            std::cout << "ConcreteHandler1 handled the request\n";
        } else if (next) {
            next->handle(request);
        }
    }
};

class ConcreteHandler2 : public Handler {
public:
    void handle(int request) override {
        if (request == 2) {
            std::cout << "ConcreteHandler2 handled the request\n";
        } else if (next) {
            next->handle(request);
        }
    }
};

int main() {
    ConcreteHandler1 handler1;
    ConcreteHandler2 handler2;

    handler1.setNext(&handler2);

    handler1.handle(1);
    handler1.handle(2);

    return 0;
}

File: labs/lab10/lab10_4.kt
open class Handler {
    var next: Handler? = null

    open fun handle(request: Int) {
        next?.handle(request)
    }
}

class ConcreteHandler1 : Handler() {
    override fun handle(request: Int) {
        if (request == 1) {
            println("ConcreteHandler1 handled the request")
        } else {
            next?.handle(request)
        }
    }
}

class ConcreteHandler2 : Handler() {
    override fun handle(request: Int) {
        if (request == 2) {
            println("ConcreteHandler2 handled the request")
        } else {
            next?.handle(request)
        }
    }
}

fun main() {
    val handler1 = ConcreteHandler1()
    val handler2 = ConcreteHandler2()

    handler1.next = handler2

    println("Request value: 1")
    handler1.handle(1)
    println("Request value: 2")
    handler1.handle(2)
}